src/validators/model.rs

```diff
@@ -241,10 +242,13 @@ impl ModelValidator {
             } else {
                 PySet::new_bound(py, [&String::from(ROOT_FIELD)])?
             };
-            force_setattr(py, self_instance, intern!(py, DUNDER_FIELDS_SET_KEY), fields_set)?;
+            force_setattr(py, self_instance, intern!(py, DUNDER_FIELDS_SET_KEY), &fields_set)?;
             force_setattr(py, self_instance, intern!(py, ROOT_FIELD), &output)?;
+            state.fields_set_count = Some(fields_set.len() + state.fields_set_count.unwrap_or(0));
```

Comment:
I think by adding in this way, when we consider a list we're summing up the models across all list items:

```python
[{'a': 1, 'b': 2}, {'a': 2, 'b': 3}]  # total of ModelA is 2, total of ModelB is 4, so ModelB wins
```

If I follow correctly, that means that stuff like how many defaulted members exist swing the outcome depending on the input length. I get the following characteristics out where the list can flip to `A1` or `A2` type based on the count of total successful fields:

```python
from pydantic import TypeAdapter, BaseModel

class A1(BaseModel):
    a: int
    b: int = 0

class A2(BaseModel):
    a: int
    c: int = 0
    d: int = 0

fits_a1_better = {'a': 0, 'b': 0}
fits_a2_better = {'a': 0, 'c': 0, 'd': 0}

validate = TypeAdapter(list[A1] | list[A2]).validate_python

print(validate([fits_a1_better, fits_a2_better]))
#> [A2(a=0, c=0, d=0), A2(a=0, c=0, d=0)]
print(validate([fits_a1_better, fits_a1_better, fits_a2_better]))
#> [A1(a=0, b=0), A1(a=0, b=0), A1(a=0, b=0)]
print(validate([fits_a1_better, fits_a1_better, fits_a2_better, fits_a2_better]))
#> [A2(a=0, c=0, d=0), A2(a=0, c=0, d=0), A2(a=0, c=0, d=0), A2(a=0, c=0, d=0)]
```

... is this ok? Unsure, but it's a less intuitive result than current behaviour, I think. This is why I think list validators might want to remove `fields_set`. And then there's a whole load of validators with similar problems 😥

Reply:
@davidhewitt,

What other validators would you also like to disable this metric for?

Reply:
Another case where `left_to_right` helps:

```py
from pydantic import TypeAdapter, BaseModel, Field
from typing import Annotated

class A1(BaseModel):
    a: int
    b: int = 0

class A2(BaseModel):
    a: int
    c: int = 0
    d: int = 0

fits_a1_better = {'a': 0, 'b': 0}
fits_a2_better = {'a': 0, 'c': 0, 'd': 0}

validate = TypeAdapter(Annotated[list[A1] | list[A2], Field(union_mode='left_to_right')]).validate_python

print(validate([fits_a1_better, fits_a2_better]))
#> [A1(a=0, b=0), A1(a=0, b=0)]
print(validate([fits_a1_better, fits_a1_better, fits_a2_better]))
#> [A1(a=0, b=0), A1(a=0, b=0), A1(a=0, b=0)]
print(validate([fits_a1_better, fits_a1_better, fits_a2_better, fits_a2_better]))
#> [A1(a=0, b=0), A1(a=0, b=0), A1(a=0, b=0), A1(a=0, b=0)]
```

Reply:
> @davidhewitt,
> 
> What other validators would you also like to disable this metric for?

I think basically every container of variable size has the same problem? sets, lists, tuples, dicts etc

Reply:
Yeah. Makes sense. I think for now we can move forward with the current fix and add a docs update for the `left_to_right` suggestions.